### PR TITLE
Feat: implements #127

### PR DESCRIPTION
**Pull request by SukkaW**
I create a draft PR since It is just a Proof of Concept for #127. The PR is meant to allow for discussion and showing some refactoring issues going on.

### CODE REVIEW

**Comment by SukkaW** [comment]
@Scrum Also, what do you think about add `tree.oncontent`, `tree.onattrs` and `tree.onnode` to posthtml?

**Comment by maltsev** [comment]
I quickly skimmed the code. I like this new approach! Let me take a more detailed look over the weekend.

@SukkaW did you do any speed benchmarks?

**Comment by SukkaW** [comment]
> @SukkaW did you do any speed benchmarks?

Not "official" and correct benchmark approach, only tested on my own website, and it is indeed faster.

It will be better if we have:

- Discuss a better API design (better naming, etc.)
- An Official API provided by PostHTML that we can utilize.

File: lib/htmlnano.es6 [hunk 1]
```diff
@@ -19,8 +23,57 @@ function htmlnano(options = {}, preset = safePreset) {
                 throw new Error('Module "' + moduleName + '" is not defined');
             }
 
-            let module = require('./modules/' + moduleName);
-            promise = promise.then(tree => module.default(tree, options, moduleOptions));
+            const module = require('./modules/' + moduleName);
+
+            if (module.once) {
+                // It is a "once" version of htmlnano module
+                if (module.onattrs) {
```
Comment (maltsev) [inline lib/htmlnano.es6:30]: Maybe go with camelCase to improve readability?

```suggestion
                if (module.onAttrs) {
```

File: lib/htmlnano.es6 [hunk 2]
```diff
@@ -19,8 +23,57 @@ function htmlnano(options = {}, preset = safePreset) {
                 throw new Error('Module "' + moduleName + '" is not defined');
             }
 
-            let module = require('./modules/' + moduleName);
-            promise = promise.then(tree => module.default(tree, options, moduleOptions));
+            const module = require('./modules/' + moduleName);
+
+            if (module.once) {
```
Comment (maltsev) [inline lib/htmlnano.es6:28]: I don't quite understand the difference between traditional and _"once"_ modules.

File: lib/htmlnano.es6 [hunk 3]
```diff
@@ -19,8 +23,57 @@ function htmlnano(options = {}, preset = safePreset) {
...
+
+                    for (const handler of nodeHandlers) {
+                        node = handler(node, node);
+                    }
+
+                    return node;
+                });
+
+                return tree;
+            });
```
Comment (maltsev) [inline lib/htmlnano.es6:76]: This block has too deep nesting. Maybe move it out of the first `if`?

```
if (attrsHandlers.length + contentsHandlers.length + nodeHandlers.length > 0) {
    return promise;
}

promise = promise.then(tree => {
...
```

**Review by maltsev — COMMENTED** [review]
> An Official API provided by PostHTML that we can utilize.

Which one? Did they add something new in the recent releases?

--- 

I really appreciate you did such a big refactoring! It's really nice having such great contributors :-)

File: lib/htmlnano.es6 (hunk 2 above)
Comment (SukkaW) [reply lib/htmlnano.es6:28]: `once` module means the module is "independent": `on*` method only requires information from the `node` itself.

Some modules (`mergeScripts`, `removeOptionalTags`) requires "context" (the node before and after, the parent node, etc.), thus it is not "independent" and can not be refactored to `once` module

Comment (folknor) [reply lib/htmlnano.es6:28]: `node`, `forEach`, `zerocontext`, `nocontext`, `nodeprocessor`, `strictnode`, `node-only`, `perNode`, `every`

**Comment by SukkaW** [comment]
@maltsev 

I pick up the PR, rewrite it, and make it ready for review.

And here are the changes after the rewrite:

- Use `camelCase` for named export: `onAttrs`, `onNode`, `onContent` instead of previous `onattrs`, `onnode` and `oncontent`.
- The new module format no longer requires a `once = true` named export.

**Comment by SukkaW** [comment]
@maltsev I forgot to mark the PR as ready for review.

**Comment by maltsev** [comment]
Thanks! 

@SukkaW Unfortunately, I don't have time in the near future to do the proper review. But all your previous contributions were of top-notch quality. So I'm fine with just accepting this PR and releasing a new htmlnano version. Are you okay with that?

**Comment by SukkaW** [comment]
> Thanks!
> 
> @SukkaW Unfortunately, I don't have time in the near future to do the proper review. But all your previous contributions were of top-notch quality. So I'm fine with just accepting this PR and releasing a new htmlnano version. Are you okay with that?

Actually, I have already tested the PR on my own project (published as [`@sukkaw/htmlnano`](https://npmjs.com/package/@sukka/htmlnano)) and it works fine. The PR shouldn't cause issues.

**Comment by maltsev** [comment]
Awesome! I'm going to merge it now then. I'll release the new version in the coming days.

**Comment by maltsev** [comment]
I released the new version.

**Pull request merged**